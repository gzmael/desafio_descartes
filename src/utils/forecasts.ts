import { Forecast } from 'models/Forecast'

export const getDayByDate = (date: string) =>
  new Intl.DateTimeFormat('pt-br', {
    day: '2-digit',
    timeZone: 'America/Fortaleza',
  }).format(new Date(date))

export const getHourByTimestamp = (time: number) =>
  new Intl.DateTimeFormat('pt-br', {
    hour: '2-digit',
    minute: '2-digit',
    timeZone: 'America/Fortaleza',
  }).format(new Date((time + 10800) * 1000))

export const getDayWeekByDate = (
  date: string,
  style?: 'short' | 'long' | 'narrow',
) =>
  new Intl.DateTimeFormat('pt-br', {
    weekday: style ?? 'long',
    timeZone: 'America/Fortaleza',
  }).format(new Date(date))

export const getWeekAndDaysFromForecasts = (forecasts: Forecast[]) => {
  const dates = [
    ...new Set(forecasts.map((forecast) => forecast.date.split(' ')[0])),
  ]
  return dates.map((date) => {
    return {
      date,
      day: getDayByDate(date + ' 12:00:00'),
      week: getDayWeekByDate(date + ' 12:00:00', 'short'),
    }
  })
}
