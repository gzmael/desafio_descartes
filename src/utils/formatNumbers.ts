export const formatTemperature = (temp: number, maxFrac = 1) => {
  return new Intl.NumberFormat('pt-br', {
    maximumFractionDigits: maxFrac,
  }).format(temp)
}

export const formatMsToKmH = (value: number, digits = 2) => {
  return new Intl.NumberFormat('pt-br', {
    maximumFractionDigits: digits,
  }).format(value * 3.6)
}
