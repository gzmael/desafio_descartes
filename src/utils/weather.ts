import { Weather, WeatherResponse } from 'models/Weathear'
import { formatMsToKmH, formatTemperature } from './formatNumbers'
import { Forecast, ForecastReponse } from 'models/Forecast'

export const getDetailListFromWeather = (weather: Weather) => {
  return [
    {
      title: 'Vento',
      value: formatMsToKmH(weather.wind) + ' km/h',
    },
    {
      title: 'Umidade',
      value: weather.humidity + ' %',
    },
    {
      title: 'Sensação',
      value: formatTemperature(weather.feels) + ' ºC',
    },
    {
      title: 'Min - Máx',
      value: `${formatTemperature(weather.min, 0)}ºC - ${formatTemperature(
        weather.max,
        0,
      )}ºC`,
    },
    {
      title: 'Visibilidade',
      value: weather.visibility / 1000 + ' km',
    },
    {
      title: 'Nuvens',
      value: weather.clouds + ' %',
    },
  ]
}

export const getWeatherFromData = (data: WeatherResponse) => {
  const { clouds, main, visibility, weather, wind, sys } = data
  return {
    clouds: clouds.all,
    description: weather[0].description,
    feels: main.feels_like,
    humidity: main.humidity,
    icon: weather[0].icon,
    max: main.temp_max,
    min: main.temp_min,
    temperature: main.temp,
    wind: wind.speed,
    pressure: main.pressure,
    sunrise: sys.sunrise,
    sunset: sys.sunset,
    visibility,
  }
}

export const getForecastsFromData = (data: ForecastReponse): Forecast[] => {
  return data.list.map((forecast) => {
    const {
      clouds,
      dt_txt: date,
      main,
      visibility,
      weather,
      wind,
      dt,
    } = forecast
    return {
      clouds: clouds.all,
      description: weather[0].description,
      feels: main.feels_like,
      humidity: main.humidity,
      icon: weather[0].icon,
      max: main.temp_max,
      min: main.temp_min,
      temperature: main.temp,
      wind: wind.speed,
      pressure: main.pressure,
      visibility,
      date,
      timestamp: dt,
    }
  })
}
