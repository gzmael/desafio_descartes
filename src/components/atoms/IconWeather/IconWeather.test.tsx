import { render, screen } from '@testing-library/react'

import { IconWeatherProps, IconWeather } from '.'
import { cn } from 'utils/cls-merge'

describe('Component/Atom/IconWeather', () => {
  it('should render IconWeather with default variant', () => {
    const props: IconWeatherProps = {
      icon: '03n',
    }

    const defaultClass = cn(['h-10 w-10'])

    const { container } = render(<IconWeather {...props} />)

    expect(screen.getByAltText(/Weather Icon/i)).toBeInTheDocument()

    expect(container.firstChild).toHaveClass(defaultClass)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render IconWeather with large variant', () => {
    const props: IconWeatherProps = {
      icon: '03n',
      size: 'lg',
    }

    const defaultClass = cn(['h-16 w-16'])

    const { container } = render(<IconWeather {...props} />)

    expect(screen.getByAltText(/Weather Icon/i)).toBeInTheDocument()

    expect(container.firstChild).toHaveClass(defaultClass)

    expect(container.firstChild).toMatchSnapshot()
  })
})
