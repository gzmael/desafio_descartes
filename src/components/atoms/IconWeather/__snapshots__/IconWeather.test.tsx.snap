// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component/Atom/IconWeather > should render IconWeather with default variant 1`] = `
<img
  alt="Weather Icon"
  class="h-10 w-10"
  src="https://openweathermap.org/img/wn/03n.png"
/>
`;

exports[`Component/Atom/IconWeather > should render IconWeather with large variant 1`] = `
<img
  alt="Weather Icon"
  class="h-16 w-16"
  src="https://openweathermap.org/img/wn/03n@2x.png"
/>
`;
